src/rust/rust-proto/src/graplinc/grapl/api/graph_query_service/v1beta1/server.rs

```diff
@@ -0,0 +1,190 @@
+    }
+}
+
+/// A server construct that drives the GraphQueryApi implementation.
+pub struct GraphQueryServiceServer<T, E>
+where
+    T: GraphQueryApi<Error = E> + Send + Sync + 'static,
+    E: Into<Status> + Send + Sync + 'static,
+{
+    server: GraphQueryServiceServerProto<T>,
+    addr: SocketAddr,
+    shutdown_rx: Receiver<()>,
+}
+
+impl<T, E> GraphQueryServiceServer<T, E>
+where
+    T: GraphQueryApi<Error = E> + Send + Sync + 'static,
+    E: Into<Status> + Send + Sync + 'static,
+{
+    pub fn builder(
+        service: T,
+        addr: SocketAddr,
+        shutdown_rx: Receiver<()>,
+    ) -> GraphQueryServiceServerBuilder<T, E> {
+        GraphQueryServiceServerBuilder::new(service, addr, shutdown_rx)
+    }
+
+    pub async fn serve(self) -> Result<(), GraphQueryServiceServerError> {
+        let (healthcheck_handle, health_service) =
+            init_health_service::<GraphQueryServiceServerProto<T>, _, _>(
+                || async { Ok(HealthcheckStatus::Serving) },
+                std::time::Duration::from_millis(500),
+            )
+            .await;
+
+        let listener = TcpListener::bind(self.addr)
+            .await
+            .map_err(GraphQueryServiceServerError::BindError)?;
+
+        Server::builder()
+            .trace_fn(|request| {
+                tracing::trace_span!(
+                    "GraphQuery",
+                    headers = ?request.headers(),
+                    method = ?request.method(),
+                    uri = %request.uri(),
+                    extensions = ?request.extensions(),
+                )
+            })
+            .add_service(health_service)
+            .add_service(self.server)
+            .serve_with_incoming_shutdown(
+                TcpListenerStream::new(listener),
+                self.shutdown_rx.map(|_| ()),
+            )
+            .then(|result| async move {
+                healthcheck_handle.abort();
+                result
+            })
+            .await?;
+        Ok(())
+    }
+}
+
+pub struct GraphQueryServiceServerBuilder<T, E>
+where
+    T: GraphQueryApi<Error = E> + Send + Sync + 'static,
+    E: Into<Status> + Send + Sync + 'static,
+{
+    server: GraphQueryServiceServerProto<T>,
+    addr: SocketAddr,
+    shutdown_rx: Receiver<()>,
+}
+
+impl<T, E> GraphQueryServiceServerBuilder<T, E>
+where
+    T: GraphQueryApi<Error = E> + Send + Sync + 'static,
+    E: Into<Status> + Send + Sync + 'static,
+{
+    /// Create a new builder for a GraphQueryServiceServer,
+    /// taking the required arguments upfront.
+    pub fn new(service: T, addr: SocketAddr, shutdown_rx: Receiver<()>) -> Self {
+        Self {
+            server: GraphQueryServiceServerProto::new(service),
+            addr,
+            shutdown_rx,
+        }
+    }
+
+    /// Consumes the builder and returns a new `GraphQueryServiceServer`.
+    /// Note: Panics on invalid build state
+    pub fn build(self) -> GraphQueryServiceServer<T, E> {
```

Comment:
I'm not clear on what this builder is actually doing... it looks like it's just another name for the server it's supposed to be building. Am I missing some magic somewhere?

Reply:
I agree that this builder doesn't follow the patterns we use elsewhere in the code and I'm tempted to rip it out for being different. @colin-grapl  ?

Reply:
It's a builder that takes the required arguments in its constructor. If we add other arguments, such as "with_timeout", they would go here. If you want to rip it out you can go for it, we'll probably need it later but I'm fine just writing it again.

Reply:
I'd rather remove it, since it's not really a "builder" at the moment. If there's cause to have a builder later, then it's easy to add. Better to not have dead / confusing / "maybe someday" code hanging around.

Reply:
eh i'll leave it be

Reply:
oh missed chris ocmment. yeah i can delete it

Reply:
this ended up being very much the right move - the server impl was pretty different from the rest of our services